build_your_own_story_book.py

```diff
@@ -35,38 +35,65 @@
 find_ending(endings, choices4, 1) => -1
 find_ending(endings, choices4, 2) => 15
 """
+from typing import List, Tuple, Dict, Set
 
 
-def find_ending(endings, choices, option):
-    curPage = 1
+def preprocess_choices(choices: List[List[int]]) -> Dict[int, Tuple[int, int]]:
+    """
+    Preprocess the choices into a dictionary for O(1) lookups.
+
+    Args:
+        choices (List[List[int]]): The list of choices, where each choice is represented as [current_page, option_1, option_2].
+
+    Returns:
+        Dict[int, Tuple[int, int]]: A dictionary mapping the current_page to a tuple of (option_1, option_2).
+    """
+    return {choice[0]: (choice[1], choice[2]) for choice in choices}
+
+
+def find_ending(endings: Set[int], choices: Dict[int, Tuple[int, int]], option: int) -> int:
+    """
+    Finds the ending page based on the choices made.
+
+    Args:
+        endings (Set[int]): A set of ending pages.
+        choices (Dict[int, Tuple[int, int]]): A dictionary of choices with pages as keys and their options as values.
+        option (int): The choice option to always take (1 or 2).
+
+    Returns:
+        int: The ending page reached, or -1 if stuck in a loop.
+    """
+    cur_page = 1
     visited = set()
-    while curPage not in endings:
-        if curPage in visited:
+
+    while cur_page not in endings:
+        if cur_page in visited:
             return -1
-        visited.add(curPage)
-        nextPage = None
-        for choice in choices:
-            if choice[0] == curPage:
-                if option == 1:
-                    nextPage = choice[1]
-                else:
-                    nextPage = choice[2]
-                    break
-        curPage = nextPage
-    return curPage
+        visited.add(cur_page)
+        if cur_page in choices:
+            cur_page = choices[cur_page][option - 1]
+        else:
+            cur_page += 1
 
+    return cur_page
+
+
+def run_tests() -> None:
+    """
+    Runs the test cases to validate the solution.
+    """
+    TEST_CASES = [
+        (set([6, 15, 21, 30]), [[3, 14, 2]], 1, 15),
+        (set([6, 15, 21, 30]), [[3, 14, 2]], 2, -1),
+        (set([6, 15, 21, 30]), [[5, 11, 28], [9, 19, 29], [14, 16, 20], [18, 7, 22], [25, 6, 30]], 1, 6),
```

Comment:
It gives `AssertionError` on this test case

Reply:
Fixed